### ISSUE TITLE
Code folding will collapse entire line containing closing bracket

### ISSUE THREAD

**Issue by moojek** [microsoft/vscode-cpptools#5521]
**Type: LanguageService**

**Describe the bug**
- OS and Version: Linux, 5.6.13-arch1-1
- VS Code Version: 1.44.2
- C/C++ Extension Version: 0.28.0
- Other extensions you installed (and if the issue persists after disabling them): github.vscode-pull-request-github james-yu.latex-workshop liviuschera.noctis Yes, issue persists
- Does this issue involve using SSH remote to run the extension on a remote machine?: Dunno how to check
- A clear and concise description of what the bug is, including information about the workspace (i.e. is the workspace a single project or multiple projects, size of the project, etc): in all types of workspaces, when you use formatting style which puts "else if"'s in the same line as closing bracket to previous conditional, folding hides too many lines (look at screenshots).

**Steps to reproduce**

1. Type this code 
```cpp
if(cond) {
//
} else {
//
}
```
2. Try to fold the "if" conditional
3. See error: "else" gets folded with the previous line

**Expected behavior**

fold is only applied to what's inside the if, not touching the "else" line

<details>
  <summary><strong>Logs</strong></summary>

Nothing shows.
</details>

**Screenshots**

Before folding: 
![image](https://user-images.githubusercontent.com/42188458/82226598-fb2c1a80-9926-11ea-9906-c89f8b0ad892.png)
Folding with extensions disabled: 
![image](https://user-images.githubusercontent.com/42188458/82226745-2adb2280-9927-11ea-8a95-aac050d537ed.png)
Folding with C++ extension:
![image](https://user-images.githubusercontent.com/42188458/82226672-15fe8f00-9927-11ea-9de6-ec5bd9f8c388.png)

**Additional context**

**Comment by Colengms** [comment]
Hi @moojek .  Thanks for reporting this.

Fixing this is tricky, due to limitations of VS Code's folding API.  It's line based, and doesn't take character/column positions into account.  Our folding algorithm is shared with VS, and can fold at character/column resolution, if that were supported  We could omit the last line from the fold, but that would instead do the wrong thing if there is content prior to the closing bracket.

Related issues that might lead to a solution from VS Code:

https://github.com/microsoft/vscode/issues/50840
https://github.com/microsoft/vscode/issues/3352

**Comment by moojek** [comment]
I mean I am pretty sure it worked a while ago. Did something change since?

**Comment by Colengms** [comment]
@moojek C++ code folding was added in v0.28.0 of the extension.  If you'd prefer to use VS Code's built-in indent-based folding, you can disable C++ code folding with the following in settings.json : `"C_Cpp.codeFolding": "Disabled"`

**Comment by moojek** [comment]
Thanks! I will use it as workaround now. What am I possibly losing (what are the advantages of C++ folding over VSC)?

**Comment by sean-mcmanus** [comment]
The main advantage of the new code folding is when region blocks are left-indented, such as
```
int main()
{
  int i;
#ifdef FOO
   int j;
#endif
}
```
in which case the "main" doesn't fold correctly.

**Comment by Colengms** [comment]
There are also other issues.  The following class cannot be collapsed entirely, only per permission section:
```
class A
{
private:
    int i;

public:
    void foo();
};
```
The folding algorithm in the C/C++ Extension understands the syntax of the language, so is better able to determine what should be grouped.

Though, as per the symptom you encountered, VS Code's folding resolution is per-line, creating some problems when that group should be started/ended within a line.

**Comment by xgdgsc** [comment]
I also see this. Especially misleading with `else if` folded.

**Comment by ClydeHobart** [comment]
Same. Similar to the `else if` case is a bracketed case statement (for case-based variable declaration):
```
switch (foo) {
    case bar: {
        int x = 0;
        ...
    } case baz: {
        int y = 0;
        ...
    }
}
```

collapses to

```
switch (foo) {
    case bar: {...
        int y = 0;
        ...
    }
}
```

**Comment by informedecommerce** [comment]
`if () {
 } else if () {
     let object  {}  
 }`

collapses to

`if () { ...
     let object = {}`

**Comment by informedecommerce** [comment]
In VS Code settings.json add this

"editor.foldingStrategy": "indentation"

**Comment by beachstrider** [comment]
Thanks Rob, your solution is really straightforward and useful.

**Comment by sean-mcmanus** [comment]
Fixed with https://github.com/microsoft/vscode-cpptools/releases/tag/1.5.0-insiders2 .

**Comment by vacing** [comment]
> Fixed with https://github.com/microsoft/vscode-cpptools/releases/tag/1.5.0-insiders2 .

it doesn't work for me when I remote connect to linux folder from windows, both are 1.5.0-insiders3.
it works fine on windows locally.

**Comment by Colengms** [comment]
Thanks @vacing .  It looks like we have more work to do here.  I opened https://github.com/microsoft/vscode-cpptools/issues/7792 to track scenarios that were not addressed by the fix associated with this issue.

**Comment by sean-mcmanus** [comment]
@vacing By "not working" I assume you mean it doesn't work with the nested if case you provided. We have a fix at https://github.com/microsoft/vscode-cpptools/pull/7793 if you want to code review and/or test the fix yourself.

**Comment by vacing** [comment]
> @vacing By "not working" I assume you mean it doesn't work with the nested if case you provided. We have a fix at #7793 if you want to code review and/or test the fix yourself.

thank you for your rapid response
but sorry, I am not familiar with vscode plugin development, even I would like to test the fix very much, I don't know how...

**Comment by sean-mcmanus** [comment]
@vacing The fix is in 1.5.0 (https://github.com/microsoft/vscode-cpptools/releases/tag/1.5.0).

**Comment by vacing** [comment]
> @vacing The fix is in 1.5.0 (https://github.com/microsoft/vscode-cpptools/releases/tag/1.5.0).

thank you, it workes fine for the previous case now, now I have a new bad case
```cpp
    if (true) {
        if (true) {
            ;
        } else {
            ;
        }
        ;
    } else if(true)  // else or else if(), the most important is no bracket here
        ;
```
